src/qibolab/unrolling.py

```diff
@@ -6,16 +6,23 @@
 from dataclasses import asdict, dataclass, field, fields
 from functools import total_ordering
 
-from .pulses import PulseSequence
+from .pulses import Pulse, PulseSequence
+from .pulses.envelope import Rectangular
 
 
 def _waveform(sequence: PulseSequence):
     # TODO: deduplicate pulses (Not yet as drivers may not support it yet)
-    # TODO: count Rectangular and delays separately (Zurich Instruments supports this)
-    # TODO: Any constant part of a pulse should be counted only once (Zurich Instruments supports this)
-    # TODO: check if readout duration is faithful for the readout pulse (I would only check the control pulses)
-    # TODO: Handle multiple qubits or do all devices have the same memory for each channel ?
-    return sequence.duration - sequence.ro_pulses.duration
+    # TODO: VirtualZ deserves a separate handling
+    # TODO: any constant part of a pulse should be counted only once (Zurich Instruments supports this)
+    # TODO: handle multiple qubits or do all devices have the same memory for each channel ?
+    return sum(
+        (
+            (pulse.duration if not isinstance(pulse.envelope, Rectangular) else 1)
+            if isinstance(pulse, Pulse)
+            else 1
```

Comment:
I am guessing this case is for `Delay` or `VirtualZ`, so it could be
```suggestion
            else 0
```
?

Reply:
The case of `VirtualZ` is different, and might be actually a 0 (I left a `TODO` for that). But for `Delay` you actually need one number: the duration.

Thus, the `Rectangular` should be two numbers, duration and amplitude. However, the truth is that at this level they're more or less random, and 1 or 0 is practically the same, since you will consume your memory first with other pulses (or you will consume a different type of memory).

We should go more in the details of the various instruments memories, and we have #698 for that. For the time being, let's keep it just a little bit less random (avoid counting non-existing samples for rectangular pulses and delays), and refine them later on.
Retrieving the `sampling_rate` from somewhere would already be a macroscopic advancement. But not required as well (since it can be traded by a different limit in the `Bounds` instance).